args <- commandArgs(TRUE)
## get arguments
####################
inputfileDir=args[1]
outputDir=args[2]
prefix=args[3]
bootstrap=args[4]
Chr=args[5]
Pops=if(is.na(args[6])){NULL}else{args[6]}
prop_chr_th=as.numeric(args[7])

## install missing dependencies
##############################
packages <- c("tidyverse", "ggplot2", "gridExtra", "stringr") 
missing_pkg <- sapply(packages, function(i) !nzchar(system.file(package = i)))

if(any(missing_pkg)){
  install.packages(packages[missing_pkg], repos='http://cran.rstudio.com/')
}

## load dependencies
####################
sapply(packages, library, character.only = T)
#library(ggpubr)

## functions
############
read_vcfStatsFiles <- function(prefix, extension, chromosomes = "all", 
                              bootstraps = "all", pops = NULL, inputfileDir = ".", ...) {

  if (identical(pops, character(0)) | is.null(pops)) { pops = "1" }
  missingF <- NULL
  df <- data.frame()
  for (pop in pops) {
    popname <- ifelse(pop=="1", "", paste0("_", pop))
    for (chr in chromosomes) {
      chrNum <- ifelse(chr == "all", "",  paste0("_", chr))
      for (b in bootstraps) {
        bNum <- paste0("_b",  b)
        filename <- paste0(prefix, chrNum, popname, bNum, extension)
        if (file.exists(file.path(inputfileDir, filename))){
          f <- read_delim(file.path(inputfileDir, filename), ...)
          f$subset = b
          f$chr = chr
          f$pop = pop
          df <- rbind(df, f)
        } else {
          missingF <- c(missingF, filename)
        }
      }
    }
  }
  df$chr <- factor(df$chr, levels = chromosomes)
  df$pop <- factor(df$pop, levels = pops)
  df$subset <- factor(df$subset, levels = bootstraps)
  if (length(missingF) == 0) {
	cat("All '", extension, "' files from", prefix, "have been read successfully.")
  } else {
	warning("Following files not found: ", sapply(missingF, function(i) paste("\n", i)), immediate. = T)
  }
  cat("\n")
  return(df)
}

###
# get number of subsets
#######################
if (bootstrap == "all") {
  files <- intersect(list.files(inputfileDir, prefix), list.files(inputfileDir, "_b[0-9]"))
  bootstraps <- sort(sub("_b", "", unique(sapply(files, function(i) str_extract(i, "_b[0-9]*")))))
} else {
  bootstraps <- as.character(bootstrap)
}
cat("number of bootstraps to analyse:", bootstraps, "\n")

###
# Get chromosomes
#################
if (Chr == "all") {
  files <- intersect(list.files(inputfileDir, prefix), list.files(inputfileDir, paste0(prefix, "_Chr[0-9]")))
  if (length(files) > 0) {
    chromosomes <- sub("_", "", str_sort(unique(sapply(files, function(i) str_extract(i, "_Chr[0-9]*"))), numeric = T))
  } else {
    chromosomes <- Chr
  }
} else {
  chromosomes <- as.character(Chr)
}
cat("Chromosomes to analyse:", chromosomes, "\n")

### 
# Find if there is Population data
##################################
# get populations
#files <- intersect(list.files(inputfileDir, prefix), list.files(inputfileDir, "\\.het"))
#Pops = unique(gsub("_Chr[0-9]*.", "", sub(prefix, "", gsub("_b[0-9].het", "", files))))
#Pops <- Pops[sapply(Pops, nchar) > 0]
if (!is.null(Pops)){
  if (file.exists(Pops)) {
    popList <- read.table(Pops)
    Pops <- unique(na.exclude(popList[,2]))
    # get number of individuals by population
    names(popList) <- c("sample", "pop")
    nInd_byPop <- na.exclude(unique(popList) %>%
                               group_by(pop) %>%
                               count())
  } else {
    Pops <- strsplit(Pops, ",")[[1]]
  }
  cat("Populations in the analysis:", Pops, "\n")
}

rm(files, bootstrap, Chr, missing_pkg, packages)
########################################################################################################################
########################################################################################################################
#########################
## Generate empty reports
#########################

## summaries report
cat(toupper("vcf basic statistics report"), "\n", paste(rep("-", nchar("vcf basic statistics report")), collapse = ""), file = file.path(outputDir, paste0(prefix, "_vcfstats_report.txt")))
cat("# This file is generated by running the script 5B_output_vcfStats.R.
# It feeds from the output of the script 5A_vcfStatistics.sh, in which vcf subsampling and population split can be performed to then, use vcftools to run various anlayses.
# This file contains the basic summaries of the following metrics:
#\t1- Variant quality: Summary statistics of variant quality grouped by chromosomes and sampling subset (if provided) obtained from '.lqual' files.
#\t2- Variant mean depth: Summary statistics of variant mean depth grouped by chromosomes and sampling subset (if provided) obtained from '.ldepth.mean' files.
#\t3- Variant missigness: Summary statistics variant missigness grouped by chromosomes and sampling subset (if provided) obtained from '.miss' files.
#\t4- Mean depth per individual: Summary statistics of mean depth per individual grouped by chromosomes and sampling subset (if provided) obtained from '.idepth' files.
#\t5- Proportion of missing data per individual: Summary statistics grouped by chromosomes and sampling subset (if provided) obtained from '.imiss' files.
#\t6- Heterozygosity and inbreeding coefficient per individual: Summary statistics of inbreeding coefficients grouped by chromosomes and sampling subset (if provided) obtained from '.het' files.
#\t7- Minor allele frequencies: Summary statistics of MAF grouped by chromosomes and sampling subset (if provided) obtained from '.frq' files.
#\t8- Hardy-Weinberg Equilibrium tests: Number of sites, number of significant of HWE exact tests p-values, number of sites exhibiting heterozygosisty deficit and number of sites exhibiting heterozygosisty excess grouped by chromosomes and sampling subset (if provided) obtained from '.hwe' files.
#- Number of vcf subsets in the analyses:", length(bootstraps), "(", bootstraps, ")
#- Number of chromosomes:", length(chromosomes), "(", chromosomes, ")
#- Number of populations:", length(Pops), "(", Pops, ")\n", file = file.path(outputDir, paste0(prefix, "_vcfstats_report.txt")))
## ind outliers report
cat(toupper("vcf individual outliers\n"), file = file.path(outputDir, paste0(prefix, "_ind_outliers_report.txt")))
## MAF variant outliers
cat(toupper("vcf MAF outliers\n"), file = file.path(outputDir, paste0(prefix, "_maf_outliers_report.txt")))
## HWE variant outliers
cat(toupper("vcf HWE outliers\n"), file = file.path(outputDir, paste0(prefix, "_hwe_outliers_report.txt")))
## parameters file
cat("", file = file.path(outputDir, paste0(prefix, "_filtering_parameters.txt")))
## remove_ind parameter file
cat("", file = file.path(outputDir, paste0(prefix, "_remove_ind.txt")))

########################################################################################################################
########################################################################################################################
# 1.Variant quality
########################################################################################################################
########################################################################################################################
cat("Reporting variant quality...\n")
var_qual <- read_vcfStatsFiles(prefix, extension = ".lqual", inputfileDir = inputfileDir,
				bootstraps = bootstraps, chromosomes = chromosomes, pops = NULL, 
				skip = 1, col_names = c("chr", "pos", "qual"), delim = "\t")

cat("Summarizing variant quality...\n")
### Summary report
##################
summ.qual <- var_qual %>%
  group_by(chr, subset) %>%
  summarise(across(qual, list(min = min,
                          median = median,
                          mean = mean,
                          stdev = sd,
                          q25 = ~quantile(., 0.25),
                          q75 = ~quantile(., 0.75),
                          max = max)))

cat("Writting variant quality summary...\n")
### Writting summary
cat(paste0("\nVariant quality\n", paste(rep("-", nchar("Variant quality")), collapse = ""), "\n"), file = file.path(outputDir, paste0(prefix, "_vcfstats_report.txt")), append = T)
write_delim(as.data.frame(summ.qual), file = file.path(outputDir, paste0(prefix, "_vcfstats_report.txt")), delim = "\t", col_names = T, quote = "none", append = T)
cat("\n",  file = file.path(outputDir, paste0(prefix, "_vcfstats_report.txt")), append = T)

### chromosome plot
###################
tab_var_qual <- as.data.frame(var_qual %>%
  group_by(chr) %>%
  count(qual) %>%
  reframe(qual = c(0,qual),
          n = c(1, 1-cumsum(n)/sum(n))))

cat("Generating plot...\n")
pl.qual <- ggplot(var_qual, aes(qual, ..scaled.., fill = as.factor(chr), col = as.factor(chr))) + 
  geom_density(alpha = 0.5, col = NA) +
  geom_line(data = tab_var_qual, aes(x = qual, y = n, group = as.factor(chr)), col = "black", alpha = 0.5, linewidth = 0.75) + 
  geom_boxplot(data = var_qual, aes(x = qual, y = -0.025), width = 0.05, fill = "grey90", col = "grey40", outlier.colour = "red", outlier.size = 0.8) +
  ggtitle("Variant quality") +
  scale_fill_viridis_d() + 
  labs(fill="chr") + 
  scale_x_continuous("Phred score") +
  scale_y_continuous("Scaled density", 
                     sec.axis = sec_axis(~.,
                                         name = "Proportion of variants kept (%)", 
                                         labels = scales::percent)) + 
  theme_light() +
  theme(text = element_text(size = 16), plot.title = element_text(hjust = 0.5), legend.position = "none")

###########################
# MINIMUM QUALITY THRESHOLD
###########################
# We recommend setting a minimum threshold of 30 and filtering more strongly on other aspects of the data.
min_qual=30

# Write value
cat(c("MIN_QUAL", min_qual, "\n"),
      file = file.path(outputDir, paste0(prefix, "_filtering_parameters.txt")), append = T)
########################################

if(diff(abs(range(var_qual$qual) - mean(var_qual$qual))) > max(var_qual$qual)/2){
  inset.pl <- ggplotGrob(ggplot(var_qual[var_qual$qual > 0,]) +
    geom_histogram(aes(x = qual), binwidth = 10, fill = "grey80", col = "grey50") +
    theme_light() +
    ggtitle("Non-zero quality variants") +
    theme(text = element_text(size = 16), plot.title = element_text(hjust = 0.5)))
  pl.qual = pl.qual + annotation_custom(grob = inset.pl, xmin=max(var_qual$qual)/2, xmax=max(var_qual$qual), ymin=0.5, ymax=1)
}

# remove 'var_qual' to avoid unnecessary use of memory
rm(var_qual, summ.qual, tab_var_qual)

########################################################################################################################
########################################################################################################################
# 2.Variant mean depth
########################################################################################################################
########################################################################################################################
cat("Reporting variant mean depth...\n")
var_depth <- read_vcfStatsFiles(prefix, extension = ".ldepth.mean", inputfileDir = inputfileDir,
                                bootstraps = bootstraps, chromosomes = chromosomes, pops = NULL,
                                skip = 1, col_names = c("chr", "pos", "mean_depth", "var_depth"), delim = "\t")

cat("Summarizing mean depth...\n")
### Summary report
##################
summ.depth <- var_depth %>%
  group_by(chr, subset) %>%
  summarise(across(mean_depth, list(min = min,
                              median = median,
                              mean = mean,
                              stdev = sd,
                              q25 = ~quantile(., 0.25),
                              q75 = ~quantile(., 0.75),
                              max = max)))

cat("Writting variant mean depth summary...\n")
### Writting summary
cat(paste0("\nVariant mean depth\n", paste(rep("-", nchar("Variant mean depth")), collapse = ""), "\n"), file = file.path(outputDir, paste0(prefix, "_vcfstats_report.txt")), append = T)
write_delim(as.data.frame(summ.depth), file = file.path(outputDir, paste0(prefix, "_vcfstats_report.txt")), delim = "\t", col_names = T, quote = "none", append = T)
cat("\n",  file = file.path(outputDir, paste0(prefix, "_vcfstats_report.txt")), append = T)

### chromosome plot
###################
tab_var_depth <- var_depth %>%
  group_by(chr) %>%
  count(mean_depth) %>%
  reframe(mean_depth = c(0,mean_depth),
          n = c(1, 1-cumsum(n)/sum(n)))

cat("Generating plot...\n")
pl.depth <- ggplot(var_depth, aes(mean_depth, ..scaled.., fill = as.factor(chr), col = as.factor(chr))) + 
  geom_density(alpha = 0.5, col = NA) +
  geom_line(data = tab_var_depth, aes(x = mean_depth, y = n, group = as.factor(chr)), col = "black", alpha = 0.5, linewidth = 0.75) + 
  geom_boxplot(data = var_depth, aes(x = mean_depth, y = -0.025), width = 0.05, fill = "grey90", col = "grey40", outlier.colour = "red", outlier.size = 0.8) +
  ggtitle("Variant mean depth") +
  labs(fill="chr") + 
  scale_fill_viridis_d() +
  scale_x_continuous("mean depth") +
  scale_y_continuous("Scaled density", 
                     sec.axis = sec_axis(~.,
                                         name = "Proportion of variants kept (%)", 
                                         labels = scales::percent)) + 
  theme_light() +
  theme(text = element_text(size = 16), plot.title = element_text(hjust = 0.5), legend.position = "none")

# We could set our minimum coverage at the 5 and 95% quantiles but we should keep in mind 
# that the more reads that cover a site, the higher confidence our basecall is. 
# 10x is a good rule of thumb as a minimum cutoff for read depth, although if we wanted to be conservative, 
# we could go with 15x.

# Set a good maximum depth cufoff is also very important.
# Regions that have extremely high coverage likely reflects mapping/assembly errors and also paralogous or repetitive regions. 
# We want to exclude these as they will bias our analyses. 
# Usually a good rule of thumb is something the mean depth x 2.

################################
### MIN and MAX DEPTH THRESHOLDS
################################
min_depth=8
max_depth <- as.integer(var_depth %>%
                     summarise(mean_depth = mean(mean_depth, na.rm = T)) * 2)

# Write value
cat(c("MIN_DEPTH", min_depth, "\n"),
      file = file.path(outputDir, paste0(prefix, "_filtering_parameters.txt")), append = T)
cat(c("MAX_DEPTH", max_depth, "\n"),
      file = file.path(outputDir, paste0(prefix, "_filtering_parameters.txt")), append = T)
########################################

pl.depth <- pl.depth + 
  geom_vline(xintercept = min_depth, linetype = "dashed", col = "grey50", linewidth = 1) +
  geom_vline(xintercept = max_depth, linetype = "dashed", col = "grey50", linewidth = 1) 

if(diff(abs(range(var_depth$mean_depth) - mean(var_depth$mean_depth))) > max(var_depth$mean_depth)/2){
  inset.pl <- ggplotGrob(ggplot(var_depth[var_depth$mean_depth < (mean(var_depth$mean_depth) + 2 * sd(var_depth$mean_depth)),]) +
                           geom_histogram(aes(x = mean_depth), binwidth = 1, fill = "grey80", col = "grey50") +
                           theme_light() +
                           theme(text = element_text(size = 16)))
  pl.depth = pl.depth + annotation_custom(grob = inset.pl, xmin=max(var_depth$mean_depth)/2, xmax=max(var_depth$mean_depth), ymin=0.5, ymax=1)
}
# remove 'var_depth' to avoid unnecessary use of memory
rm(var_depth, summ.depth, tab_var_depth)

########################################################################################################################
########################################################################################################################
# 3.Variant missingness
########################################################################################################################
########################################################################################################################
cat("Reporting variant missingness...\n")
var_miss <- read_vcfStatsFiles(prefix, extension = ".lmiss", inputfileDir = inputfileDir,
                                bootstraps = bootstraps, chromosomes = chromosomes, pops = NULL,
                                skip = 1, col_names = c("chr", "pos", "nchr", "nfiltered", "nmiss", "fmiss"), delim = "\t")

cat("Summarizing missingness...\n")
## Summary report
#################
summ.miss <- var_miss %>%
  group_by(chr, subset) %>%
  summarise(across(fmiss, list(min = min,
                                    median = median,
                                    mean = mean,
                                    stdev = sd,
                                    q25 = ~quantile(., 0.25),
                                    q75 = ~quantile(., 0.75),
                                    max = max)))

cat("Writting variant missingness summary...\n")
### Writting summary
cat(paste0("\nVariant missingness\n", paste(rep("-", nchar("Variant missingness")), collapse = ""), "\n"), file = file.path(outputDir, paste0(prefix, "_vcfstats_report.txt")), append = T)
write_delim(as.data.frame(summ.miss), file = file.path(outputDir, paste0(prefix, "_vcfstats_report.txt")), delim = "\t", col_names = T, quote = "none", append = T)
cat("\n",  file = file.path(outputDir, paste0(prefix, "_vcfstats_report.txt")), append = T)

### chromosome plot
###################
tab_var_miss <- var_miss %>%
  group_by(chr) %>%
  count(fmiss) %>%
  reframe(fmiss = c(0,fmiss, 1),
          n = c(0, cumsum(n)/sum(n), 1))

cat("Generating plot...\n")
pl.miss <- ggplot(var_miss, aes(fmiss, ..scaled..,fill = as.factor(chr), col = as.factor(chr))) + 
  geom_density(alpha = 0.5, col = NA) +
  geom_line(data = tab_var_miss, aes(x = fmiss, y = n, group = as.factor(chr)), col = "black", alpha = 0.5, linewidth = 0.75) + 
  ggtitle("Variant missingness") +  
  geom_boxplot(data = var_miss, aes(x = fmiss, y = -0.025), width = 0.05, fill = "grey90", col = "grey40", outlier.colour = "red", outlier.size = 0.8) +
  labs(fill="chr") + 
  scale_fill_viridis_d() +
  scale_x_continuous("proportion of missingness") +
  scale_y_continuous("Scaled density", 
                     sec.axis = sec_axis(~.,
                                         name = "Proportion of variants kept (%)", 
                                         labels = scales::percent)) + 
  theme_light() +
  theme(text = element_text(size = 16), plot.title = element_text(hjust = 0.5), legend.position = "bottom")

# We will remove all sites where over 10% of individuals are missing a genotype. 
# One thing to note here is that vcftools inverts the direction of missigness, 
# so our 10% threshold means we will tolerate 90% missingness (yes this is confusing and counterintuitive… but that’s the way it is!). 
# Typically missingness of 75-95% is used.

########################################
### PROPORTION OF MISSING DATA THRESHOLD
########################################
miss_threshold=0.90

cat(c("MISS_THRESHOLD", miss_threshold, "\n"),
      file = file.path(outputDir, paste0(prefix, "_filtering_parameters.txt")), append = T)
########################################
pl.miss <- pl.miss + 
  geom_vline(xintercept = 1 - miss_threshold, linetype = "dashed", col = "grey50", linewidth = 1) 

# remove 'var_miss' to avoid unnecessary use of memory
rm(var_miss, summ.miss, tab_var_miss)

#################################
## Save Variant-related plots
#################################
cat("Saving variant-based plot...\n")
pl.varstat <- grid.arrange(pl.qual, pl.depth, pl.miss, nrow = 3)  
#pl.varstat <- ggarrange(pl.qual, pl.depth, pl.miss, labels = letters[1:3], common.legend = T, nrow = 2, legend = "bottom") + bgcolor("white")
ggsave(file.path(outputDir, paste0(prefix, "_variant_based_statistics.png")), plot = pl.varstat, height = 20, width = 10)

rm(pl.varstat, pl.qual, pl.depth)
########################################################################################################################
########################################################################################################################
# 4.Mean depth per individual
########################################################################################################################
########################################################################################################################
cat("Reporting depth per individual...\n")
ind_depth <- read_vcfStatsFiles(prefix, extension = ".idepth", inputfileDir = inputfileDir,
                                bootstraps = bootstraps, chromosomes = chromosomes, pops = NULL,
                                skip = 1, col_names =  c("ind", "nsites", "depth"), delim = "\t")

cat("Summarizing depth per individual...\n")
## Summary report
#################
summ.ind_depth <- ind_depth %>%
  group_by(chr, subset) %>%
  summarise(across(depth, list(min = min,
                             median = median,
                             mean = mean,
                             stdev = sd,
                             q25 = ~quantile(., 0.25),
                             q75 = ~quantile(., 0.75),
                             max = max)))

cat("Writting individual mean depth summary...\n")
### Writting summary
cat(paste0("\nIndividual mean depth\n", paste(rep("-", nchar("Individual mean depth")), collapse = ""), "\n"), file = file.path(outputDir, paste0(prefix, "_vcfstats_report.txt")), append = T)
write_delim(as.data.frame(summ.ind_depth), file = file.path(outputDir, paste0(prefix, "_vcfstats_report.txt")), delim = "\t", col_names = T, quote = "none", append = T)
cat("\n",  file = file.path(outputDir, paste0(prefix, "_vcfstats_report.txt")), append = T)

## Outliers detection
#####################
ind_depth_outliers_perChr <- sapply(as.character(unique(ind_depth$chr)),
                                           function(chr) ind_depth$ind[which(ind_depth$chr == chr
                                                                             & ind_depth$depth[ind_depth$chr == chr]
                                                                             %in% boxplot.stats(ind_depth$depth[ind_depth$chr == chr])$out)])

ind_depth_outliers_perChr <- ind_depth_outliers_perChr[sapply(ind_depth_outliers_perChr, length) > 0]

tab_ind_depth_outliers <- table(unlist(ind_depth_outliers_perChr))/length(chromosomes)

# Write individual depth outliers per chromosome
############
cat(paste0("\nIndividual depth outliers per chromosome\n", paste(rep("-", nchar("Individual depth outliers per chromosome")), collapse = ""), "\n"), 
				file = file.path(outputDir, paste0(prefix, "_ind_outliers_report.txt")), append = T)
if (length(ind_depth_outliers_perChr) > 0){
  for (l in 1:length(ind_depth_outliers_perChr)){
    cat(names(ind_depth_outliers_perChr)[[l]], " : ", ind_depth_outliers_perChr[[l]], 
        file = file.path(outputDir, paste0(prefix, "_ind_outliers_report.txt")),
        append = T)
    cat("\n",  file = file.path(outputDir, paste0(prefix, "_ind_outliers_report.txt")), append = T)
  }
}
cat("\n",  file = file.path(outputDir, paste0(prefix, "_ind_outliers_report.txt")), append = T)

######################
## INDIVIDUAL OUTLIERS
######################
ind_depth_outliers <- names(tab_ind_depth_outliers)[tab_ind_depth_outliers > prop_chr_th]
######################

### chromosome plot
###################
tab_ind_depth <- ind_depth %>%
  group_by(chr) %>%
  count(depth) %>%
  reframe(depth = c(0,depth,max(depth)),
          n = c(0, cumsum(n)/sum(n), 1))

cat("Generating plot...\n")
pl.ind_depth <- ggplot(ind_depth, aes(depth, fill = as.factor(chr), col = as.factor(chr))) +
  geom_histogram(binwidth = 1, alpha = 0.5,  position = "identity", col = NA) +
  geom_vline(xintercept = min_depth, linetype = "dashed", col = "grey50", linewidth = 1) +
  geom_line(data = tab_ind_depth, aes(x = depth, y = n*max(table(round(ind_depth$depth), ind_depth$chr)), group = as.factor(chr)), col = "black", alpha = 0.5, linewidth = 0.75) +
  geom_boxplot(data = ind_depth, aes(x = depth, y = -0.025), width = 0.05, fill = "grey90", col = "grey40", outlier.colour = "red", outlier.size = 0.8) +
  ggtitle("Mean depth per individual") +
  labs(fill="chr") +
  scale_fill_viridis_d() +
  scale_x_continuous("mean depth") +
  scale_y_continuous("Counts", 
                     sec.axis = sec_axis(~./max(table(round(ind_depth$depth), ind_depth$chr)),
                                         name = "Proportion of individuals kept (%)", 
                                         labels = scales::percent)) + 
  theme_light() +
  theme(text = element_text(size = 16), plot.title = element_text(hjust = 0.5), legend.position = "none")

nInd_byPop <- unique(ind_depth[,c("ind", "pop")]) %>%
  group_by(pop) %>%
  count()

# remove 'ind_depth' to avoid unnecessary use of memory
rm(summ.ind_depth, ind_depth_outliers_perChr, tab_ind_depth_outliers)

########################################################################################################################
########################################################################################################################
# 5.Proportion of missing data per individual
########################################################################################################################
########################################################################################################################
cat("Reporting missing data per individual...\n")
ind_miss <- read_vcfStatsFiles(prefix, extension = ".imiss", inputfileDir = inputfileDir,
                                bootstraps = bootstraps, chromosomes = chromosomes, pops = NULL,
                                skip = 1, col_names =  c("ind", "ndata", "nfiltered", "nmiss", "fmiss"), delim = "\t")

cat("Summarizing missing data per individual...\n")
## Summary report
#################
summ.ind_miss <- ind_miss %>%
  group_by(chr, subset) %>%
  summarise(across(fmiss, list(min = min,
                               median = median,
                               mean = mean,
                               stdev = sd,
                               q25 = ~quantile(., 0.25),
                               q75 = ~quantile(., 0.75),
                               max = max)))

cat("Writting individual missing data summary...\n")
### Writting summary
cat(paste0("\nIndividual missing data\n", paste(rep("-", nchar("Individual missing data")), collapse = ""), "\n"), file = file.path(outputDir, paste0(prefix, "_vcfstats_report.txt")), append = T)
write_delim(as.data.frame(summ.ind_miss), file = file.path(outputDir, paste0(prefix, "_vcfstats_report.txt")), delim = "\t", col_names = T, quote = "none", append = T)
cat("\n",  file = file.path(outputDir, paste0(prefix, "_vcfstats_report.txt")), append = T)

## Outliers detection
#####################
ind_miss_outliers_perChr <- sapply(as.character(unique(ind_miss$chr)),
                                           function(chr) ind_miss$ind[which(ind_miss$chr == chr
                                                                             & ind_miss$fmiss[ind_miss$chr == chr]
                                                                             %in% boxplot.stats(ind_miss$fmiss[ind_miss$chr == chr])$out)])

ind_miss_outliers_perChr <- ind_miss_outliers_perChr[sapply(ind_miss_outliers_perChr, length) > 0]

tab_ind_miss_outliers <- table(unlist(ind_miss_outliers_perChr))/length(chromosomes)

# Write individual missing data outliers per chromosome
############
cat(paste0("\nIndividual missing data outliers per chromosome\n", paste(rep("-", nchar("Individual missing data outliers per chromosome")), collapse = ""), "\n"), 
                                file = file.path(outputDir, paste0(prefix, "_ind_outliers_report.txt")), append = T)
if (length(ind_miss_outliers_perChr) > 0){
  for (l in 1:length(ind_miss_outliers_perChr)){
    cat(names(ind_miss_outliers_perChr)[[l]], " : ", ind_miss_outliers_perChr[[l]], 
        file = file.path(outputDir, paste0(prefix, "_ind_outliers_report.txt")),
        append = T)
    cat("\n",  file = file.path(outputDir, paste0(prefix, "_ind_outliers_report.txt")), append = T)
  }
}
cat("\n",  file = file.path(outputDir, paste0(prefix, "_ind_outliers_report.txt")), append = T)

######################
## INDIVIDUAL OUTLIERS
######################
ind_miss_outliers <- names(tab_ind_miss_outliers)[tab_ind_miss_outliers > prop_chr_th]
######################

### chromosome plot
###################
tab_ind_miss <- ind_miss %>%
  group_by(chr) %>%
  count(fmiss) %>%
  reframe(fmiss = c(0, fmiss, 1),
          n = c(0,  cumsum(n)/sum(n), 1))

cat("Generating plot...\n")
pl.ind_miss <- ggplot(ind_miss, aes(fmiss, ..scaled..,fill = as.factor(chr), col = as.factor(chr))) +
  geom_density(alpha = 0.5, col = NA) +
  geom_vline(xintercept = 1 - miss_threshold, linetype = "dashed", col = "grey50", linewidth = 1) +
  geom_line(data = tab_ind_miss, aes(x = fmiss, y = n, group = as.factor(chr)), col = "black", alpha = 0.5, linewidth = 0.75) +
  geom_boxplot(data = ind_miss, aes(x = fmiss, y = -0.025), width = 0.05, fill = "grey90", col = "grey40", outlier.colour = "red", outlier.size = 0.8) +
  ggtitle("Missing data") +
  labs(fill="chr") +
  scale_fill_viridis_d() +
  scale_x_continuous("proprotion of missing data") +
  scale_y_continuous("Scaled density", 
                     sec.axis = sec_axis(~.,
                                         name = "Proportion of individuals kept (%)", 
                                         labels = scales::percent)) + 
  theme_light() +
  theme(text = element_text(size = 16), plot.title = element_text(hjust = 0.5), legend.position = "bottom")

# remove 'ind_miss' to avoid unnecessary use of memory
rm(summ.ind_miss, ind_miss_outliers_perChr, tab_ind_miss_outliers)

#################################
## Save Individual-related plots
#################################
pl.indstat <- grid.arrange(pl.ind_depth, pl.ind_miss)  
#pl.indstat <- ggarrange(pl.ind_depth, pl.ind_miss, labels = letters[1:2], common.legend = T, nrow = 2, legend = "bottom") + bgcolor("white")

ggsave(file.path(outputDir,  paste0(prefix, "_individual_based_statistics.png")), plot = pl.indstat, height = 15, width = 10)

rm(pl.indstat, pl.ind_depth, pl.ind_miss, ind_depth, ind_miss, tab_ind_depth, tab_ind_miss)
gc()
########################################################################################################################
########################################################################################################################
# 6.Heterozygosity and inbreeding coefficient per individual
########################################################################################################################
########################################################################################################################
cat("Reporting inbreeding coefficient per individual...\n")

# generate loop to iterate through populations as otherwise it crashes due to out-of-memory issue.

cat(paste0("\nInbreeding coefficient per individual\n", paste(rep("-", nchar("Inbreeding coefficient per individual")), collapse = ""), "\n"), file = file.path(outputDir, paste0(prefix, "_vcfstats_report.txt")), append =T)
cat(paste0("\nIndividual heterozygosity outliers per chromosome\n", paste(rep("-", nchar("Individual heterozygosity outliers per chromosome")), collapse = ""), "\n"),
    file = file.path(outputDir, paste0(prefix, "_ind_outliers_report.txt")), append = T)
col_names = T
ind_het_df = data.frame()
ind_het_outliers = NULL

for (pop in Pops) { 

  cat("Reading", prefix, "files from", pop, "population...\n")
  ind_het <- read_vcfStatsFiles(prefix, extension = ".het", inputfileDir = inputfileDir,
                                  bootstraps = bootstraps, chromosomes = chromosomes, pops = pop,
                                  skip = 1, col_names = c("ind","ho", "he", "nsites", "f"), delim = "\t")
  
  cat("Summarizing inbreeding coefficient per individual...\n")
  ## Summary report
  #################
  summ.ind_het <- ind_het %>%
    group_by(pop, ind, chr) %>%
    summarise(across(f, list(min = min,
                                 median = median,
                                 mean = mean,
                                 stdev = sd,
                                 q25 = ~quantile(., 0.25),
                                 q75 = ~quantile(., 0.75),
                                 max = max)))
  
  ### Writting summary
  write_delim(as.data.frame(summ.ind_het), file = file.path(outputDir, paste0(prefix, "_vcfstats_report.txt")), delim = "\t", col_names = col_names, quote = "none", append = T)
  col_names = F
  
  ind_het_df <- rbind(ind_het_df, ind_het %>%
					group_by(pop, ind, chr) %>%
					summarise(f = mean(f, na.rm = TRUE)))
  
  ## Outliers detection
  ##################### 
  ind_het_outliers_perChr <- sapply(chromosomes, 
                                            function(chr) ind_het_df$ind[which(ind_het_df$chr == chr 
                                                                             & ind_het_df$f[ind_het_df$chr == chr] 
                                                                             %in% boxplot.stats(ind_het_df$f[ind_het_df$chr == chr])$out)])
  
  ind_het_outliers_perChr <- ind_het_outliers_perChr[sapply(ind_het_outliers_perChr, length) > 0]
  
  tab_ind_het_outliers <- table(unlist(ind_het_outliers_perChr))/length(chromosomes)
  
  # Write individual heterozygosity outliers per chromosome
  ############
  cat(paste0("\n", pop, "\n", paste(rep("-", nchar(pop)), collapse = ""), "\n"),
      file = file.path(outputDir, paste0(prefix, "_ind_outliers_report.txt")), append = T)
  if (length(ind_het_outliers_perChr) > 0){
    for (l in 1:length(ind_het_outliers_perChr)){
      cat(names(ind_het_outliers_perChr)[[l]], " : ", ind_het_outliers_perChr[[l]], 
          file = file.path(outputDir, paste0(prefix, "_ind_outliers_report.txt")),
                   append = T)
      cat("\n",  file = file.path(outputDir, paste0(prefix, "_ind_outliers_report.txt")), append = T)
    }
  }
  cat("\n",  file = file.path(outputDir, paste0(prefix, "_ind_outliers_report.txt")), append = T)
  
  ######################
  ## INDIVIDUAL OUTLIERS
  ######################
  ind_het_outliers <- c(ind_het_outliers, names(tab_ind_het_outliers)[tab_ind_het_outliers > prop_chr_th])
  ######################
  
  rm(ind_het, summ.ind_het, ind_het_outliers_perChr, tab_ind_het_outliers)
}
cat("\n",  file = file.path(outputDir, paste0(prefix, "_vcfstats_report.txt")), append = T)
#### END FOR LOOP

### Writing all individual outliers
remove_ind <- unique(c(ind_depth_outliers, ind_miss_outliers, ind_het_outliers))
cat(remove_ind, file = file.path(outputDir, paste0(prefix, "_remove_ind.txt")))

tab_ind_het <- ind_het_df %>%
    group_by(pop, chr) %>%
    count(f) %>%
    reframe(f = c(min(f), f, max(f)),
            n = c(0, cumsum(n)/sum(n), 1))

cat("Generating Inbreeding coefficient by population plot...\n")
xlims <- c(-ceiling(max(abs(ind_het_df$f))*100)/100, ceiling(max(abs(ind_het_df$f))*100)/100)

### Population plot
###################
pl.ind_het_pop <- ggplot(ind_het_df, aes(f, ..scaled..,fill = pop, col = pop)) +
  geom_density(alpha = 0.5, col = NA) +
  #geom_step(data = tab_ind_het, aes(x = f, y = n, group = pop), linewidth = 0.75) +
  geom_boxplot(data = ind_het_df, aes(x = f, y = -0.025), width = 0.05, fill = "grey90", col = "grey40", outlier.colour = "red", outlier.size = 0.8) +
  ggtitle("Inbreeding coefficient by population") +
  labs(fill="pop") + guides(colour = "none") + 
  scale_color_viridis_d() +
  scale_fill_viridis_d() +
  scale_x_continuous("Inbreeding coefficient", limits = xlims) +
  scale_y_continuous("Scaled density", 
                     sec.axis = sec_axis(~.,
                                         name = "Proportion of individuals kept (%)", 
                                         labels = scales::percent)) + 
  theme_light() +
  theme(text = element_text(size = 16), plot.title = element_text(hjust = 0.5)) + 
  facet_wrap(.~as.factor(chr)) + theme(legend.position = "bottom")

cat("Generating Inbreeding coefficient by chromosome plot...\n")
### chromosome plot
###################
pl.ind_het_chr <- ggplot(ind_het_df, aes(f, ..scaled..,fill = as.factor(chr), col = as.factor(chr))) + 
  geom_density(alpha = 0.5, col = NA) +
  #geom_step(data = tab_ind_het, aes(x = f, y = n, group = as.factor(chr)), col = "black", alpha = 0.5, linewidth = 0.75) + 
  geom_boxplot(data = ind_het_df, aes(x = f, y = -0.025), width = 0.05, fill = "grey90", col = "grey40", outlier.colour = "red", outlier.size = 0.8) +
  ggtitle("Inbreeding coefficient by chromosome") +
  labs(fill="chr") + 
  scale_fill_viridis_d() +
  scale_x_continuous("Inbreeding coefficient", limits = xlims) +
  scale_y_continuous("Scaled density", 
                     sec.axis = sec_axis(~.,
                                         name = "Proportion of individuals kept (%)", 
                                         labels = scales::percent)) + 
  theme_light() +
  theme(text = element_text(size = 16), plot.title = element_text(hjust = 0.5)) + 
  facet_wrap(.~pop) + theme(legend.position = "bottom")

# Save plots...
cat("Saving plots...\n")
ggsave(file.path(outputDir, paste0(prefix, "_individual_het_bypop.png")), plot = pl.ind_het_chr, height = 10, width = 10)
ggsave(file.path(outputDir, paste0(prefix, "_individual_het_bychr.png")), plot = pl.ind_het_pop, height = 10, width = 10)

# check Hardy-Weinberg equilibrium (applies within populations)
# H-W assumptions:
#   - no mutation
#   - random mating
#   - no gene flow
#   - infinite population size
#   - no selection.
# check Wahlund-effect
# high levels of allelic dropout (strongly negative F) or DNA contamination (highly positive F).

# get number of individuals by population
nInd_byPop <- unique(ind_het_df[, c("ind", "pop")]) %>%
  group_by(pop) %>%
  count()

rm(ind_het_df, tab_ind_het, pl.ind_het_chr, pl.ind_het_pop)
gc()
########################################################################################################################
########################################################################################################################
# 7.Minor allele frequency
########################################################################################################################
########################################################################################################################
cat("Reporting Minor allele frequency...\n")

# generate loop to iterate through populations as otherwise it crashes due to out-of-memory issue.

cat(paste0("\nMinor allele frequency\n", paste(rep("-", nchar("Minor allele frequency")), collapse = ""), "\n"), 
    file = file.path(outputDir, paste0(prefix, "_vcfstats_report.txt")), append = T)
col_names = T
var_freq_df = data.frame()

cat("Computing MAF threshold...\n")
min_allele_occurrence = 2 # for diploid organisms
maf_threshold=setNames(round(min_allele_occurrence / (nInd_byPop$n * 2), 2), nInd_byPop$pop)

for (pop in Pops) {
  
  cat("Reading", pop, "population...\n")
  
  for (chr in chromosomes) {
    cat("Reading", chr, "chromosome...\n")
    var_freq <- read_vcfStatsFiles(prefix, extension = ".frq", inputfileDir = inputfileDir,
                                  bootstraps = bootstraps, chromosomes = chr, pops = pop,
                                  skip = 1, col_names = c("chr", "pos", "nalleles", "nchr", "a1", "a2"), delim = "\t")
    if (nrow(var_freq) > 0){
	# find minor allele frequency
	var_freq$maf <- var_freq %>% dplyr::select(a1, a2) %>% apply(1, function(z) min(z))
	var_freq <- unique(var_freq[,c("pop", "chr", "pos", "maf")])
	var_freq <- var_freq[,c("pop", "chr", "maf")]
	
	cat("Summarizing MAF...\n")
    ## Summary report
    #################
	summ.maf <- var_freq %>%
		group_by(pop, chr) %>%	
		summarise(across(maf, list(min = min,
                                 median = median,
                                 mean = mean,
                                 stdev = sd,
                                 q25 = ~quantile(., 0.25),
                                 q75 = ~quantile(., 0.75),
                                 max = max)))
    
    ### Writting summary
	write_delim(as.data.frame(summ.maf), file = file.path(outputDir, paste0(prefix, "_vcfstats_report.txt")), 
                delim = "\t", col_names = col_names, quote = "none", append = T)
	col_names = F
    
	var_freq_df <- rbind(var_freq_df, var_freq[var_freq$maf > 0,])
    
	rm(var_freq, summ.maf)
    }
    ########################
    ### OUTLIER MAF VARIANTS  
    ########################
    maf_otulier_pos <- var_freq_df$pos[var_freq_df$maf < maf_threshold[pop]]
    ## Write outliers to report file  
    cat(paste0("\n\t- ", pop, "\n\t", paste(rep("-", nchar(paste("-", pop))), collapse = ""), "\n"), 
        file = file.path(outputDir, paste0(prefix, "_maf_outliers_report.txt")), append = T)
    cat(maf_otulier_pos, file = file.path(outputDir, paste0(prefix, "_maf_outliers_report.txt")), append = T)
    cat("\n", file = file.path(outputDir, paste0(prefix, "_maf_outliers_report.txt")), append = T)

  }
}
cat("\n",  file = file.path(outputDir, paste0(prefix, "_vcfstats_report.txt")), append = T)
# END FOOR LOOP

### MIN ALLELE FREQUENCY by pop
###############################
# MAF = min_allele_occurrence / n_individuals * 2 
# Setting MAF cutoffs is actually not that easy or straightforward. 
# Hard MAF filtering (i.e. setting a high threshold) can severely bias estimation of 
# the site frequency spectrum and cause problems with demographic analyses. 
# Similarly, an excesss of low frequency,  ^`^xsingleton ^`^y SNPs (i.e. only occurring in one individual) 
# can mean you keep many uniformative loci in your dataset that make it hard to model things like population structure.

# It is best practice to produce one dataset with a good MAF threshold 
# and keep another without any MAF filtering at all.

##################################
### MIN ALLELE FREQUENCY THRESHOLD
##################################
for (i in 1:length(maf_threshold)) cat(paste0("MAF_", names(maf_threshold)[i]), maf_threshold[i], "\n", 
                                       file = file.path(outputDir, paste0(prefix, "_filtering_parameters.txt")), append = T)
#cat(c("POP", as.character(nInd_byPop$pop), "\n"),
#      file = file.path(outputDir, paste0(prefix, "_filtering_parameters.txt")), append = T)
#cat(c("MAF", maf_threshold, "\n"),
#      file = file.path(outputDir, paste0(prefix, "_filtering_parameters.txt")), append = T)
##################################

tab_var_freq <- var_freq_df %>%
    group_by(pop, chr) %>%
    count(maf) %>%
    reframe(maf = c(0, maf, 0.5),
            n = c(0, cumsum(n)/sum(n), 1))

cat("Generating plots...\n")
### Population plot
###################

pl.maf_pop <- ggplot(var_freq_df, aes(maf, ..scaled..,fill = pop, col = pop)) +
    geom_density(alpha = 0.5, col = NA) +
    geom_step(data = tab_var_freq, aes(x = maf, y = n, group = pop), linewidth = 0.75) +
    geom_boxplot(aes(x = (maf), y = -0.025), width = 0.05, fill = "grey90", col = "grey40", outlier.colour = "red", outlier.size = 0.8) +
    ggtitle("MAF distribution by population") +
    labs(fill="pop") +
    scale_color_viridis_d() +
    scale_fill_viridis_d() +
    scale_x_continuous("MAF") +
    scale_y_continuous("Scaled density", 
                     sec.axis = sec_axis(~.,
                                         name = "Proportion of variants kept (%)", 
                                         labels = scales::percent)) + 
    theme_light() +
    theme(text = element_text(size = 16), plot.title = element_text(hjust = 0.5)) + 
    facet_wrap(.~chr) + theme(legend.position = "bottom")


cat("Saving variant_maf_bychr plot...\n")
ggsave(file.path(outputDir, paste0(prefix, "_variant_maf_bychr.png")), plot = pl.maf_pop, height = 15, width = 15)
rm(pl.maf_pop)

### chromosome plot
###################
pl.maf_chr <- ggplot(var_freq_df, aes(maf, ..scaled..,fill = as.factor(chr), col = as.factor(chr))) +
  geom_density(alpha = 0.5, col = NA) +
  geom_step(data = tab_var_freq, aes(x = maf, y = n, group = as.factor(chr)),  col = "black", linewidth = 0.75) +
  geom_boxplot(aes(x = maf, y = -0.025), width = 0.05, fill = "grey90", col = "grey40", outlier.colour = "red", outlier.size = 0.8) +
  geom_vline(data = data.frame(pop = names(maf_threshold), maf = (maf_threshold)), aes(xintercept = maf), 
             linetype = "dashed", linewidth = 0.75, alpha = 0.5) + 
  ggtitle("MAF distribution by chromosome") +
  labs(fill="chr") +
  scale_fill_viridis_d() +
  scale_x_continuous("MAF") +
  scale_y_continuous("Scaled density", 
                     sec.axis = sec_axis(~.,
                                         name = "Proportion of variants kept (%)", 
                                         labels = scales::percent)) + 
  theme_light() +
  theme(text = element_text(size = 16), plot.title = element_text(hjust = 0.5)) +
  facet_wrap(.~pop) + theme(legend.position = "bottom")

rm(var_freq_df, tab_var_freq)

cat("Saving variant_maf_bypop plot...\n")
ggsave(file.path(outputDir, paste0(prefix, "_variant_maf_bypop.png")), plot = pl.maf_chr, height = 15, width = 15)
rm(pl.maf_chr)
gc()
########################################################################################################################
########################################################################################################################
# 8. Hardy-Weinberg Equilibrium tests
########################################################################################################################
########################################################################################################################
cat("Reporting Hardy-Weinberg Equilibrium tests...\n")

cat(paste0("\nHardy-Weinberg Equilibrium tests\n", paste(rep("-", nchar("Hardy-Weinberg Equilibrium tests")), collapse = ""), "\n"), 
    file = file.path(outputDir, paste0(prefix, "_vcfstats_report.txt")), append = T)
cat(toupper("vcf HWE outliers\n"), file = file.path(outputDir, paste0(prefix, "_hwe_outliers_report.txt")))
col_names = T
var_hwe_df = data.frame()
  
for (pop in Pops) {
  cat("Reading", pop, "population...\n")
  
  hwe_deficit_posList <- NULL
  hwe_excess_posList <- NULL
  ## Write to report file  
  cat("Writting ", pop, " HWE outliers info in ", outputDir, " directory...\n")
  cat(paste0("\n", pop, "\n", paste(rep("-", nchar(pop)), collapse = ""), "\n"), 
      file = file.path(outputDir, paste0(prefix, "_hwe_outliers_report.txt")), append = T)
  
  for (chr in chromosomes) {
    cat("Reading", chr, "chromosome...\n")
    var_hwe <- read_vcfStatsFiles(prefix, extension = ".hwe", inputfileDir = inputfileDir,
                               bootstraps = bootstraps, chromosomes = chr, pops = pop,  skip = 1, 
                               col_names = c("Chr", "pos", "observed", "expected", "ChiSq_HWE", "P_HWE",      
                                             "p_het_deficit", "p_het_excess", "subset", "chr", "pop"))
    if (nrow(var_hwe) > 0){
  
		var_hwe <- unique(var_hwe)
		var_hwe <- var_hwe[, c("pop", "chr", "pos", "P_HWE", "p_het_deficit", "p_het_excess")]
		cat("Summarizing HWE tests...\n")
    ## Summary report
    #################
		summ.hwe <- var_hwe %>%
			group_by(pop, chr) %>%
			summarize(n_sites = n(),
				n_significant = sum(P_HWE < 0.05),
				n_deficit = sum(p_het_deficit < 0.05),
				n_excess = sum(p_het_excess < 0.05))
    
    ### Writting summary
		write_delim(as.data.frame(summ.hwe), file = file.path(outputDir, paste0(prefix, "_vcfstats_report.txt")), delim = "\t", col_names = col_names, quote = "none", append = T)
		col_names = F
    
    ########################
    ### VARIANTS OUTSIDE HWE
    ########################
		hwe_deficit_posList <- c(hwe_deficit_posList, var_hwe$pos[var_hwe$p_het_deficit < 0.05])	
		hwe_excess_posList <- c(hwe_excess_posList, var_hwe$pos[var_hwe$p_het_excess < 0.05])
    
		var_hwe <- var_hwe[, c("pop", "chr", "P_HWE")]
    
		var_hwe_df <- rbind(var_hwe_df, var_hwe)
		rm(var_hwe, summ.hwe)
     }
  }
  ## Write outliers to report file  
  cat(paste0("\n\t- heterozygosity deficit\n\t", paste(rep("-", nchar("- heterozygosity deficit")), collapse = ""), "\n"), 
      file = file.path(outputDir, paste0(prefix, "_hwe_outliers_report.txt")), append = T)
  cat(hwe_deficit_posList, file = file.path(outputDir, paste0(prefix, "_hwe_outliers_report.txt")), append = T)
  cat("\n", file = file.path(outputDir, paste0(prefix, "_hwe_outliers_report.txt")), append = T)
  cat(paste0("\n\t- heterozygosity excess\n\t", paste(rep("-", nchar("- heterozygosity excess")), collapse = ""), "\n"), 
      file = file.path(outputDir, paste0(prefix, "_hwe_outliers_report.txt")), append = T)
  cat(hwe_excess_posList, file = file.path(outputDir, paste0(prefix, "_hwe_outliers_report.txt")), append = T)
  cat("\n", file = file.path(outputDir, paste0(prefix, "_hwe_outliers_report.txt")), append = T)
  ##################################
  ##################################
  rm(hwe_deficit_posList, hwe_excess_posList)
  ##################################
}
cat("\n",  file = file.path(outputDir, paste0(prefix, "_vcfstats_report.txt")), append = T)
# END FOOR LOOP
cat("Generating plots...\n")
### Population plot
###################

pl.hwe_pop <- ggplot(var_hwe_df, aes(P_HWE, fill = pop)) +
  geom_histogram( bins = 10, alpha = 0.5,  position = "identity") +
  geom_vline(xintercept = as.numeric(0.05), col = "red", linetype = "dashed", linewidth = 0.5) +
  ggtitle("HWE p-values distribution by population") +
  labs(fill="pop") +
  scale_color_viridis_d() +
  scale_fill_viridis_d() +
  scale_x_continuous("HWE exact test p-value") +
  scale_y_log10("Frequency", labels = function(x) format(x, scientific = TRUE)) + 
  theme_light() +
  theme(text = element_text(size = 16), plot.title = element_text(hjust = 0.5)) +
  facet_wrap(.~chr) + theme(legend.position = "bottom") 

cat("Saving variant_hwe_bychr plot...\n")
ggsave(file.path(outputDir, paste0(prefix, "_variant_hwe_bychr.png")), plot = pl.hwe_pop, height = 10, width = 10)
rm(pl.hwe_pop)

### chromosome plot
###################
pl.hwe_chr <- ggplot(var_hwe_df, aes(P_HWE, fill = as.factor(chr))) +
  geom_histogram( bins = 10, alpha = 0.5,  position = "identity") +
  geom_vline(xintercept = as.numeric(0.05), col = "red", linetype = "dashed", linewidth = 0.5) +
  ggtitle("HWE p-values distribution by chromosome") +
  labs(fill="chr") +
  scale_color_viridis_d() +
  scale_fill_viridis_d() +
  scale_x_continuous("WE exact test p-value") +
  scale_y_log10("Frequency", labels = function(x) format(x, scientific = TRUE)) + 
  theme_light() +
  theme(text = element_text(size = 16), plot.title = element_text(hjust = 0.5)) +
  facet_wrap(.~pop) + theme(legend.position = "bottom")

rm(var_hwe_df)
gc()
cat("Saving variant_hwe_bypop plot...\n")
ggsave(file.path(outputDir, paste0(prefix, "_variant_hwe_bypop.png")), plot = pl.hwe_chr, height = 10, width = 10)
rm(pl.hwe_chr)
cat(prefix, " vcfStats report has been finished.\n")
########################################################################################################################
########################################################################################################################
